### ISSUE TITLE
Backward memory usage of the embedded bag

### ISSUE THREAD

**Issue by 666easyfuture** [pytorch/pytorch#109282]
### 🚀 The feature, motivation and pitch

Here's a piece of my test code. I'd like to ask a technical question about the backward of the embedded_bag.
`import torch
import torch.nn as nn
import random
import torch.optim as optim
vocab_size = 50000000
embedding_dim = 1000
profiler_s = torch.profiler.profile(
            activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA
        ],
        #on_trace_ready=torch.profiler.tensorboard_trace_handler(log_tensorboard),
        record_shapes=True,
        profile_memory=True,
        with_stack=True,
        #use_cuda=use_gpu
        )
embedding = nn.EmbeddingBag(vocab_size, embedding_dim, mode='sum')
optimizer = optim.SGD(embedding.parameters(), lr=0.01)

input = torch.LongTensor([random.randint(0, i+1) for i in range(5000)]) 
offsets = torch.LongTensor([i*50 for i in range(100)])  
output = embedding(input, offsets)
loss = torch.randn(output.size())
output.backward(loss)
optimizer.step()`
When the program is not started, my memory displays the following
![image](https://github.com/pytorch/pytorch/assets/73371299/66140297-342b-4323-b40d-9ff45851a238)
memory = 50000000 * 1000 * 4 = 186.26GB
After the embeddingbag table is created, the memory is as follows:
![image](https://github.com/pytorch/pytorch/assets/73371299/82231163-841f-479c-b116-385d953166bd)
After the gradient matrix is calculated and before the optimizer is updated, the memory display is as follows:
![image](https://github.com/pytorch/pytorch/assets/73371299/afc3f8da-b55b-493e-ac39-27bbf618a070)

The code in \aten\src\ATen\native\EmbeddingBag.cpp is as follows:
`  auto index_grad_weight =
      at::zeros({num_weights, grad.sizes()[1]}, grad.options());`
I believe it's this large memory space created here, but in real training, batch processing is used, which means I create a matrix space as large as the weight of embeddingbag for each batch,Even though I didn't use all the indexes every time.I don't think this is very good, is this a point that can be optimized?

### Alternatives

First of all, I think only the index needed to be updated, and I just need to create a gradient matrix of the same size and store the index, and the addition process can be like the search process in the embedding_bag forward process, looping the gradient matrix and adding it to the weight matrix. Plus prefetching.it's just an idea.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by albanD** [comment]
Hi,

We use github issues only for bugs or feature requests.
Please use the forum to ask questions: https://discuss.pytorch.org/

For your particular question, the `sparse=` kwarg for the Module is here if you want sparse gradients to be generated as you seem to do: https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding_bag.html?highlight=embedding_bag#torch.nn.functional.embedding_bag
